daml-lf/spec/daml-lf-1.rst

```diff
@@ -1653,7 +1686,101 @@ bound by pattern.
     —————————————————————————————————————————————————————————————————————— MatchFail
        v 'matches' p  ⇝  Fail
 
+       
+Value equality
+~~~~~~~~~~~~~~
+
+We define here the relation ``~ᵥ`` on values that is used as equality
+check. This is a partial equivalence relation over all values, but a
+(total) equivalence relation over serialized values. This relation
+will always be used to compare values of same types::
 
+                                  ┌────────┐
+  Generic Equivalence Relation    │ v ~ᵥ w │
+                                  └────────┘
+	   
+   LitNumeric₁ and LitNumeric₂ have the same scale and value
+  ——————————————————————————————————————————————————— GenEqLitNumeric
+   LitNumeric₁ ~ᵥ LitNumeric₂
+
+   t₁ and t₂ have the same value
+  ——————————————————————————————————————————————————— GenEqLitText
+   t₁ ~ᵥ t₂
+
+   LitDate₁ and LitDate₂ have the same value
+  ——————————————————————————————————————————————————— GenEqLitDate
+   LitDate₁ ~ᵥ LitDate₂
+
+   LitTimestamp₁ and LitTimestamp₂ have the same value
+  ——————————————————————————————————————————————————— GenEqLitTimestamp
+   LitTimestamp₁ ~ᵥ LitTimestamp₂
+
+   cid₁ and cid₂ are the same
+  ——————————————————————————————————————————————————— GenEqLitContractId
+   cid₁ ~ᵥ cid₂
+
+  ——————————————————————————————————————————————————— GenEqUnit
+   () ~ᵥ ()
+
+  ——————————————————————————————————————————————————— GenEqTrue
+   'True' ~ᵥ 'True'
+
+  ——————————————————————————————————————————————————— GenEqFalse
+   'False' ~ᵥ 'False'
+
+  ——————————————————————————————————————————————————— GenEqListNil
+   'Nil' @τ₁ ~ᵥ 'Nil' @τ₂
+
+   vₕ ~ᵥ wₕ  vₜ ~ᵥ wₜ
+  ——————————————————————————————————————————————————— GenEqListCons
+   'Cons' @τ vₕ vₜ  ~ᵥ 'Cons' @τ wₜ wₜ
+
+  ——————————————————————————————————————————————————— GenEqOptionNone
+   'None' @τ ~ᵥ 'None' @σ
+
+   v ~ᵥ w
+  ——————————————————————————————————————————————————— GenEqOptionSome
+   'Some' @τ v ~ᵥ 'Some' @τ w
+
+   v₁ ~ᵥ w₁     …       vₘ ~ᵥ wₘ
+  ——————————————————————————————————————————————————— GenEqRecCon
+  Mod:T @τ₁ … @τₙ { f₁ = v₁, …, fₘ = wₘ }
+    ~ᵥ Mod:T @σ₁ … @σₙ { f₁ = w₁, …, fₘ = wₘ }
+
+   v ~ᵥ w
+  ——————————————————————————————————————————————————— GenEqVariantCon
+   Mod:T:V @τ₁ … @τₙ v ~ᵥ Mod:T:V @σ₁ … @σₙ w
+
+  ——————————————————————————————————————————————————— GenEqEnumCon
+   Mod:T:E ~ᵥ Mod:T:E
+
+   v₁ ~ᵥ w₁     …       vₙ ~ᵥ wₘ
+  ——————————————————————————————————————————————————— GenEqStructCon
+   ⟨ f₁ = v₁, …, fₘ = vₘ ⟩ ~ᵥ ⟨ f₁ = w₁, …, fₘ = wₘ ⟩
+    
+    v₁ ~ᵥ w₁     …       vₙ ~ᵥ wₙ
+  ——————————————————————————————————————————————————— GenEqTextMap
+   [ t₁ ↦ w₁; …; tₘ ↦ wₘ ]
+     ~ᵥ  [t₁ ↦ w₁; …;x tₘ ↦ wₘ]
+
+  ——————————————————————————————————————————————————— GenEqEmptyGenMap
+   〚 〛 ~ᵥ 〚 〛
+
+    i ∈ { 1 … m }      vᵢ ~ᵥ v₁'      wᵢ ~ᵥ w₁'
```

Comment:
I think it would be good to point out explicitly that this is sensitive to insertion order. This seems like something that might trip people up.

Reply:
The generic equality for genmap should not be sensitive to insertion order. This is what we agreed before and this is how equality for genmaps has been implemented in general.

This rule here is not correct.

Reply:
A fixed version of the rule might be the following three statements in first-order logic (all three are necessary, although the third could be combined with either of the first two):

- forall i in 1 .. m  there exists a unique j in 1 .. n such that v_i ~ v'_j
- forall j in 1 .. n there exists a unique i in 1 .. m such that v_i ~ v'_j
- forall i in 1 .. m, forall j in 1 .. n, if v_i ~ v'_j then w_i ~ w'_j

The "unique" in "exists a unique" prevents the consideration of maps that have repeated keys, which is fine.

Reply:
Ah, I see the rule here does *not sensitive* to insertion order, which is good. It's not very easy to see. (-:

Reply:
Oh thanks, I got confused by all the indices.

Reply:
I had a comment to underline the underline the equality of generic map is not sensitive to order of entries.